src/language-js/print/assignment.js

```diff
@@ -435,6 +442,22 @@ function getTypeArgumentsFromCallExpression(node) {
   return (node.typeParameters ?? node.typeArguments)?.params;
 }
 
+function shouldBreakBeforeConditionalType(node) {
+  function isGeneric(subNode) {
+    switch (subNode.type) {
+      case "FunctionTypeAnnotation":
+      case "GenericTypeAnnotation":
+      case "TSFunctionType":
+      case "TSTypeReference":
+        return Boolean(subNode.typeParameters);
+      default:
+        return false;
+    }
+  }
+
+  return isGeneric(node.checkType) || isGeneric(node.extendsType);
+}
```

Comment:
There is a `hasObject` in src/language-js/print/type-annotation.js seems doing the same thing, do you think we can use the same logic?

Reply:
@fisker 
It doesn't look exactly same. Do you mean
```diff
 const hasObject = node.types.some(
   (node) =>
     node.type === "ObjectTypeAnnotation" ||
     node.type === "TSTypeLiteral" ||
     // This is a bit aggressive but captures Array<{x}>
-    node.type === "GenericTypeAnnotation" ||
-    node.type === "TSTypeReference",
+    isGeneric(node)
 )

```
can be better for src/language-js/print/type-annotation.js?

Reply:
I suppose `isGeneric` should capture Array<{x}> less aggressively, though I don't understand `shouldHugType` function that contains `hasObject` yet.

Reply:
I tried https://github.com/prettier/prettier/pull/15811#discussion_r1428807026 locally. It doesn't affect any existing test cases. It seems to be slightly better.
```ts
// Prettier stable 
type MyUnion =
  VeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLong | null

// use isGeneric() in hasObect()
type MyUnion =
  | VeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryVeryLong
  | null;
```

Reply:
Sorry, didn't read carefully. Never mind.